clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -11611,6 +11611,10 @@ def warn_sycl_implicit_decl
               "declaration for a kernel type name; your program may not "
               "be portable">,
       InGroup<SyclStrict>, ShowInSystemHeader, DefaultIgnore;
+def warn_sycl_potentially_invalid_as_cast : Warning<
+  "explicit cast from %0 to %1 potentially leads to an invalid address space"
+  " cast in the resulting code">, InGroup<SyclStrict>,
+  ShowInSystemHeader;
```

Comment:
SYCL 2020 defines an `address_space_cast` function -- can we point developers to that, instead?  The definition is not easy to find in the specification, so pasting it here:

> ```c++
> template <access::address_space Space,
>           access::decorated DecorateAddress,
>           typename ElementType>
>     multi_ptr<ElementType, Space, DecorateAddress> address_space_cast(ElementType* pointer)
> ```
> Global function to create a `multi_ptr` instance from `pointer`, using the address space and decoration specified via the `Space` and `DecorateAddress` template arguments.
> 
> An implementation must return `nullptr` if the run-time value of `pointer` is not compatible with `Space`, and must issue a compile-time diagnostic if the deduced address space for `pointer` is not compatible with `Space`.

Reply:
I don't think that this thing is a solution. In fact, for me it seems like yet another place where we will emit incorrect address space casts when we implement it (I couldn't grep it in the headers).
It accepts the pointer in default address space. In our implementation it is always allowed to cast any pointer to default address space. `multi_ptr` contains a pointer to a named address space inside. That means at some point the argument pointer with default address space will be converted to a pointer with named address space, and if there was mismatch, the error will come only from SPIR-V translator or runtime, without source info, this is what people complain about. I don't know what is the compile-time diagnostic the spec mentions, but to really track the address spaces it will require some complicated analysis that is usually not implemented on FE level (and implemented by static analyzers instead).

Reply:
I didn't realize that we hadn't implemented it yet.  But putting that aside for now, I see `address_space_cast` as a solution because developers have to explicitly opt into it, and because there aren't any unsafe casts.

Unlike the `Private = (__private int *)Default;` example in your first post, `address_space_cast<access::address_space::private>(Default)` will return `nullptr` if the address space doesn't match.  I expect the implementation will call the SPIR-V [`OpGenericCastToPtrExplicit`](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#OpGenericCastToPtrExplicit) instruction, so it won't look like a regular cast.  So neither the SPIR-V translator nor runtime should throw an error -- it's the user's responsibility to check if the cast worked before using the result, to avoid dereferencing a null pointer.  The intent is to allow developers to write code that accepts a generic pointer and which dispatches (at runtime) to address-space-specific code.

We can ignore the bit about deducing address spaces.  That text only applies to SYCL implementations that cannot represent the generic address space (see [here](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:inferredAddressSpace)).

Reply:
> I expect the implementation will call the SPIR-V [OpGenericCastToPtrExplicit](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#OpGenericCastToPtrExplicit) instruction, so it won't look like a regular cast.

Yeah, that sounds reasonable. Thanks for the explanation.

The thing is, we don't have `address_space_cast` implementation right now, and we still have interface of `multi_ptr` class that contains explicit casts between address spaces. People are complaining about address space cast errors that come without source info for a long time, so I would prefer to implement the warning while we can't provide a good solution and I'm fine to remove it once the problem is resolved.

Reply:
Yeah, I agree that having the warning is going to be really helpful.  And I think we should actually keep it once `address_space_cast` exists (but modify it to point developers towards the "safe" cast, instead).

But since we can't do anything until `address_space_cast` is implemented, I'm going to approve this PR and open a feature request.  Sorry for holding things up!